src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -19,38 +19,101 @@ internal class ManagedIdentityClient
+        {
+            string identityEndpoint = EnvironmentVariables.IdentityEndpoint;
+            string identityHeader = EnvironmentVariables.IdentityHeader;
+            string identityServerThumbprint = EnvironmentVariables.IdentityServerThumbprint;
+            string msiSecret = EnvironmentVariables.IdentityHeader;
+            string msiEndpoint = EnvironmentVariables.MsiEndpoint;
+            string msiSecretMachineLearning = EnvironmentVariables.MsiSecret;
+            string imdsEndpoint = EnvironmentVariables.ImdsEndpoint;
+
+            var logger = requestContext.ServiceBundle.ApplicationLogger;
+            logger.Info("[Managed Identity] Detecting managed identity source...");
+
+            if (!string.IsNullOrEmpty(identityEndpoint) && !string.IsNullOrEmpty(identityHeader))
+            {
+                if (!string.IsNullOrEmpty(identityServerThumbprint))
+                {
+                    logger.Info("[Managed Identity] Service Fabric detected.");
+                    s_sourceName = ManagedIdentitySource.ServiceFabric;
+                }
+                else
+                {
+                    logger.Info("[Managed Identity] App Service detected.");
+                    s_sourceName = ManagedIdentitySource.AppService;
+                }
+            }
+            else if (!string.IsNullOrEmpty(msiSecretMachineLearning) && !string.IsNullOrEmpty(msiEndpoint))
+            {
+                logger.Info("[Managed Identity] Machine Learning detected.");
+                s_sourceName = ManagedIdentitySource.MachineLearning;
+            }
+            else if (!string.IsNullOrEmpty(msiEndpoint))
+            {
+                logger.Info("[Managed Identity] Cloud Shell detected.");
+                s_sourceName = ManagedIdentitySource.CloudShell;
+            }
+            else if (ValidateAzureArcEnvironment(identityEndpoint, imdsEndpoint, logger))
+            {
+                logger.Info("[Managed Identity] Azure Arc detected.");
+                s_sourceName = ManagedIdentitySource.AzureArc;
+            }
+            else if (await ImdsV2ManagedIdentitySource.GetCsrMetadataAsync(requestContext).ConfigureAwait(false))
+            {
+                logger.Info("[Managed Identity] ImdsV2 detected.");
+                s_sourceName = ManagedIdentitySource.ImdsV2;
+            }
+            else
+            {
+                s_sourceName = ManagedIdentitySource.DefaultToImds;
+            }
+
+            return s_sourceName;
+        }
+
         // Detect managed identity source based on the availability of environment variables.
         // The result of this method is not cached because reading environment variables is cheap. 
         // This method is perf sensitive any changes should be benchmarked.
+        [Obsolete("Use GetManagedIdentitySourceAsync(RequestContext) instead.")]
         internal static ManagedIdentitySource GetManagedIdentitySource(ILoggerAdapter logger = null)
```

Comment:
Logic from here needs to be consolidated with `GetManagedIdentitySourceAsync`

Reply:
resolved in refactor